### ISSUE TITLE
non-ASCII characters in the URL breaks the decoder

### ISSUE THREAD

**Issue by agenbite** [cardi/proofpoint-url-decoder#7]
I love this tool and it has become super handy. There's an email
it cannot process, however. I've traced the problem down to this
url:

```
<https://urldefense.com/v3/__https://es.linkedin.com/in/pastora-mart**Anez-samper-3049818__;w60!!D9dNQwwGXtA!VCWZpmS3jIZRe5xmjoL0_MX9ZUWwcPFyz6CnsBeDuRVdXGZbS1iTRYZ19iLt7VCJmmzf528Jpli6tlesUnQ$ >
```

Proofpoint-url-decoder gives me:

```
Traceback (most recent call last):
  File "/home/me/src/proofpoint-url-decoder/decode_email.py", line 388, in <module>
    e_clean = process_text(e)
              ^^^^^^^^^^^^^^^
  File "/home/me/src/proofpoint-url-decoder/decode_email.py", line 354, in process_text
    e_clean = re.sub(
              ^^^^^^^
  File "/usr/lib/python3.11/re/__init__.py", line 185, in sub
    return _compile(pattern, flags).sub(repl, string, count)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/me/src/proofpoint-url-decoder/decode_email.py", line 357, in <lambda>
    decode(match.group()) if "urldefense" in match.group() else match.group()
    ^^^^^^^^^^^^^^^^^^^^^
  File "/home/me/src/proofpoint-url-decoder/decode_email.py", line 295, in decode
    cleaned_url = decode_ppv3(mangled_url, unquote_url)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/me/src/proofpoint-url-decoder/decode_email.py", line 243, in decode_ppv3
    replacement_chars.append(replacement_list.pop(0))
                             ^^^^^^^^^^^^^^^^^^^^^^^
IndexError: pop from empty list
```

What do you think? It is indeed a weird url, since my bash
terminal does not recognise it if I paste it. The "**A" bit of the
url corresponds to the accented char "í", so maybe that's the
problem...

**Comment by cardi** [comment]
@agenbite great find, thank you for reporting this issue!

It looks like I did not account for handling the full set UTF-8 characters. I've tested a fix to this, and I should be able to update the scripts later today, after some testing.

Some details, if you are (or anyone else is) interested: the markers in the URL (i.e., `**A`) specify how many **bytes** (and not characters or list entries) should be replaced from the base64-encoded replacement string (`w60`). In this particular case, `**A` equals `2`, and while the replacement string `í` has length 1, it takes 2 bytes to encode `í` with UTF-8 (UTF-8 [encodes to 1-4 bytes](https://en.wikipedia.org/wiki/UTF-8#Encoding)).

The bug, then, was that I coded the URL markers to be the length of the string replacement and not the number of bytes, which led to popping from an empty list. The fix (at the moment) is build the replacement character mapping using the number of bytes by popping the character, calculating the number of bytes it is encoded to in UTF-8, and increment a counter by its size in bytes rather than string length.
